plugins/rest/rest_auth.go

```diff
@@ -122,6 +104,112 @@ func (ap *bearerAuthPlugin) Prepare(req *http.Request) error {
+	ExpiresIn   int64  `json:"expires_in"`
+}
+
+// oauth2ClientCredentialsAuthPlugin represents authentication via a bearer token in the HTTP Authorization header
+// obtained through the OAuth2 client credentials flow
+type oauth2ClientCredentialsAuthPlugin struct {
+	TokenURL     string    `json:"token_url"`
+	ClientID     string    `json:"client_id"`
+	ClientSecret string    `json:"client_secret"`
+	Scopes       *[]string `json:"scopes,omitempty"`
+
+	tokenCache    *oauth2Token
+	tlsSkipVerify bool
+}
+
+type oauth2Token struct {
+	Token     string
+	ExpiresAt time.Time
+}
+
+func (ap *oauth2ClientCredentialsAuthPlugin) NewClient(c Config) (*http.Client, error) {
+	t, err := defaultTLSConfig(c)
+	if err != nil {
+		return nil, err
+	}
+
+	// Inherit skip verify from the "parent" settings. Should this be configurable on the credentials too?
+	ap.tlsSkipVerify = c.AllowInsureTLS
+
+	if !strings.HasPrefix(ap.TokenURL, "https://") {
+		return nil, errors.New("token_url required to use https scheme")
+	}
+	if ap.ClientID == "" || ap.ClientSecret == "" {
+		return nil, errors.New("client_id and client_secret required")
+	}
+	if ap.Scopes == nil {
+		ap.Scopes = &[]string{}
+	}
+
+	return defaultRoundTripperClient(t, *c.ResponseHeaderTimeoutSeconds), nil
+}
+
+// requestToken tries to obtain an access token using the client credentials flow
+// https://tools.ietf.org/html/rfc6749#section-4.4
+func (ap *oauth2ClientCredentialsAuthPlugin) requestToken() (*oauth2Token, error) {
+	body := url.Values{"grant_type": []string{"client_credentials"}}
+	if len(*ap.Scopes) > 0 {
+		body["scope"] = []string{strings.Join(*ap.Scopes, " ")}
+	}
+
+	r, err := http.NewRequest("POST", ap.TokenURL, strings.NewReader(body.Encode()))
+	if err != nil {
+		return nil, err
+	}
+	r.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	r.SetBasicAuth(ap.ClientID, ap.ClientSecret)
+
+	client := defaultRoundTripperClient(&tls.Config{InsecureSkipVerify: ap.tlsSkipVerify}, 10)
+	response, err := client.Do(r)
+	if err != nil {
+		return nil, err
+	}
+
+	bodyRaw, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if response.StatusCode != 200 {
+		return nil, fmt.Errorf("error in response from OAuth2 token endpoint: %v", string(bodyRaw))
+	}
+
+	var tokenResponse tokenEndpointResponse
+	err = json.Unmarshal(bodyRaw, &tokenResponse)
+	if err != nil {
+		return nil, err
+	}
+
+	if strings.ToLower(tokenResponse.TokenType) != "bearer" {
+		return nil, errors.New("unknown token type returned from token endpoint")
+	}
+
+	return &oauth2Token{
+		Token:     strings.TrimSpace(tokenResponse.AccessToken),
+		ExpiresAt: time.Now().Add(time.Duration(tokenResponse.ExpiresIn) * time.Second),
+	}, nil
+}
+
+func (ap *oauth2ClientCredentialsAuthPlugin) Prepare(req *http.Request) error {
+	minTokenLifetime := float64(10)
+	if ap.tokenCache == nil || ap.tokenCache.ExpiresAt.Sub(time.Now()).Seconds() < minTokenLifetime {
```

Comment:
[nit] We could use the helper `time.Until`:

```go
time.Until(ap.tokenCache.ExpiresAt).Seconds()
```
instead of
```go
ap.tokenCache.ExpiresAt.Sub(time.Now()).Seconds()
```
😉 (Not important at all, though.)

Reply:
Thanks! Will keep that in mind for future use :)